### ISSUE TITLE
Custom training loss with custom gradients

### ISSUE THREAD

**Issue by rohanmohapatra** [tensorflow/tensorflow#36677]
I am trying to write a custom loss in Tensorflow v2, for simplicity let's say that I'm using Mean Squared Error loss as follows,
```
loss_object = tf.keras.losses.MeanSquaredError()


def loss(model, x, y, training):
  # training=training is needed only if there are layers with different
  # behavior during training versus inference (e.g. Dropout).
  y_ = model(x, training=training)
  return loss_object(y_true=y, y_pred=y_)
```

Now I know that Tensorflow does [automatic differentiation](https://stackoverflow.com/questions/48219296/why-doesnt-keras-need-the-gradient-of-a-custom-loss-function).

But I want to specify my **custom gradient**, in the BackPropagation algorithm,
if we use MSE, we have to do the following 

<a href="https://www.codecogs.com/eqnedit.php?latex=\frac{\partial&space;E}{\partial&space;w}&space;=&space;\frac{\partial&space;E}{\partial&space;y_i}&space;\frac{\partial&space;y_i}{\partial&space;net_i}&space;\frac{\partial&space;net_i}{\partial&space;w}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\frac{\partial&space;E}{\partial&space;w}&space;=&space;\frac{\partial&space;E}{\partial&space;y_i}&space;\frac{\partial&space;y_i}{\partial&space;net_i}&space;\frac{\partial&space;net_i}{\partial&space;w}" title="\frac{\partial E}{\partial w} = \frac{\partial E}{\partial y_i} \frac{\partial y_i}{\partial net_i} \frac{\partial net_i}{\partial w}" /></a>

<img src="https://latex.codecogs.com/gif.latex?\frac{\partial&space;E}{\partial&space;w}&space;=&space;(y&space;-&space;\hat{y})*&space;D(activation)&space;*&space;x" title="\frac{\partial E}{\partial w} = (y - \hat{y})* D(activation) * x" />

Is it possible in **Keras** to replace <a href="https://www.codecogs.com/eqnedit.php?latex=(y&space;-&space;\hat{y})" target="_blank"><img src="https://latex.codecogs.com/gif.latex?(y&space;-&space;\hat{y})" title="(y - \hat{y})" /></a>  with <a href="https://www.codecogs.com/eqnedit.php?latex=(y&space;-&space;p)" target="_blank"><img src="https://latex.codecogs.com/gif.latex?(y&space;-&space;p)" title="(y - p)" /></a> where `p` is a tensor that is passed during training before applying gradients.

**Comment by rohanmohapatra** [comment]
@gadagashwini can you take a look at this.

**Comment by Saduf2019** [comment]
@rohanmohapatra Could you please provide with an implementable stand alone code so we could help you resolve the issue.

**Comment by rohanmohapatra** [comment]
@Saduf2019 Hi, what I'm interested is how is the gradient of the loss w.r.t to the `pred_y` calculated? I want to replace the that gradient with my custom gradient value.

**Comment by gadagashwini** [comment]
@rohanmohapatra, Tensorflow provides two ways  to add custom loss with Keras. Please refer [this doc](https://www.tensorflow.org/guide/keras/train_and_evaluate#custom_losses). Thanks!

**Comment by rohanmohapatra** [comment]
@gadagashwini Can I also change the value of the backprop gradient for the loss? I do not want to explicitly use the AutoGrad.

**Comment by karinazad** [comment]
I have the same issue - is there a way to define a custom gradient when using keras API?

**Comment by adhilcodes** [comment]
I would like to improve it, Do anyone please help to fix it

**Comment by rmothukuru** [comment]
@rohanmohapatra,
Sorry for the delayed response. Can you please look into the Documentation of [Custom Gradients](https://www.tensorflow.org/api_docs/python/tf/custom_gradient) and let us know if it helps? Thanks!
